From 7f90e45dd7cf0d18ad423ff9850b961e9830b2d8 Mon Sep 17 00:00:00 2001
From: Bastien Nocera <hadess@hadess.net>
Date: Tue, 1 Feb 2022 14:26:02 +0100
Subject: [PATCH 1/2] gtk: Add a way to queue redrawing the base GTK widget

This will be used to request a redraw of the GTK widget should the
display be changed using properties not directly handled by the base GTK
widget, but by one of its descendants.
---
 .../gst-plugins-good/ext/gtk/gtkgstbasewidget.c  | 16 ++++++++++++++++
 .../gst-plugins-good/ext/gtk/gtkgstbasewidget.h  |  1 +
 2 files changed, 17 insertions(+)

diff --git a/subprojects/gst-plugins-good/ext/gtk/gtkgstbasewidget.c b/subprojects/gst-plugins-good/ext/gtk/gtkgstbasewidget.c
index db2c7b52d0..d3ba7e8704 100644
--- a/subprojects/gst-plugins-good/ext/gtk/gtkgstbasewidget.c
+++ b/subprojects/gst-plugins-good/ext/gtk/gtkgstbasewidget.c
@@ -593,3 +593,19 @@ gtk_gst_base_widget_set_buffer (GtkGstBaseWidget * widget, GstBuffer * buffer)
 
   GTK_GST_BASE_WIDGET_UNLOCK (widget);
 }
+
+void
+gtk_gst_base_widget_queue_draw (GtkGstBaseWidget * widget)
+{
+  /* As we have no type, this is better then no check */
+  g_return_if_fail (GTK_IS_WIDGET (widget));
+
+  GTK_GST_BASE_WIDGET_LOCK (widget);
+
+  if (!widget->draw_id) {
+    widget->draw_id = g_idle_add_full (G_PRIORITY_DEFAULT,
+        (GSourceFunc) _queue_draw, widget, NULL);
+  }
+
+  GTK_GST_BASE_WIDGET_UNLOCK (widget);
+}
diff --git a/subprojects/gst-plugins-good/ext/gtk/gtkgstbasewidget.h b/subprojects/gst-plugins-good/ext/gtk/gtkgstbasewidget.h
index 6967560795..cc957bf13b 100644
--- a/subprojects/gst-plugins-good/ext/gtk/gtkgstbasewidget.h
+++ b/subprojects/gst-plugins-good/ext/gtk/gtkgstbasewidget.h
@@ -91,6 +91,7 @@ void            gtk_gst_base_widget_finalize             (GObject * object);
 /* API */
 gboolean        gtk_gst_base_widget_set_format           (GtkGstBaseWidget * widget, GstVideoInfo * v_info);
 void            gtk_gst_base_widget_set_buffer           (GtkGstBaseWidget * widget, GstBuffer * buffer);
+void            gtk_gst_base_widget_queue_draw           (GtkGstBaseWidget * widget);
 void            gtk_gst_base_widget_set_element          (GtkGstBaseWidget * widget, GstElement * element);
 void            gtk_gst_base_widget_display_size_to_stream_size (GtkGstBaseWidget * base_widget,
                                                                  gdouble x, gdouble y,
-- 
2.34.1


From fcbcb89d492d5f09d26e6cf3d2456fc7df977ae2 Mon Sep 17 00:00:00 2001
From: Bastien Nocera <hadess@hadess.net>
Date: Tue, 1 Feb 2022 14:28:24 +0100
Subject: [PATCH 2/2] gtk: Fix rotation not being applied when paused

The video wouldn't be redrawn immediately when a rotation was applied
but the pipeline was paused, as no new buffers were scheduled to be
displayed.
---
 subprojects/gst-plugins-good/ext/gtk/gtkgstglwidget.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/subprojects/gst-plugins-good/ext/gtk/gtkgstglwidget.c b/subprojects/gst-plugins-good/ext/gtk/gtkgstglwidget.c
index 1f244002b6..b13569772a 100644
--- a/subprojects/gst-plugins-good/ext/gtk/gtkgstglwidget.c
+++ b/subprojects/gst-plugins-good/ext/gtk/gtkgstglwidget.c
@@ -752,6 +752,8 @@ gtk_gst_gl_widget_set_rotate_method (GtkGstGLWidget * gst_widget,
     priv->current_rotate_method = method;
   }
   GTK_GST_BASE_WIDGET_UNLOCK (gst_widget);
+
+  gtk_gst_base_widget_queue_draw (GTK_GST_BASE_WIDGET (gst_widget));
 }
 
 GstVideoOrientationMethod
-- 
2.34.1

